simulator/infra/replacement/cache_replacement.cpp

```diff
@@ -84,10 +85,110 @@ void LRUCacheInfo::erase_lru_element() {
     lru_hash.erase( lru_elem);
 }
 
+class Pseudo_LRUCacheInfo : public CacheReplacementInterface
+{
+    public:
+        explicit Pseudo_LRUCacheInfo( std::size_t ways);
+        void touch( std::size_t way) override;
+        std::size_t update() override;
+        std::size_t get_ways() const override { return ways; }
+
+        void set_to_erase( std::size_t ) override;
+        void allocate( std::size_t ) override;
+
+    private:
+        enum flags { LEFT, RIGHT};
+
+        tree<std::size_t> lru_tree;
+        void construct_tree( tree<std::size_t>::iterator node_it, int depth);
+        void construct_leaf_layer();
+        int calculate_depth() const;
+        std::size_t reverse_flag( std::size_t flag);
+
+        const std::size_t ways;
+        size_t leaf_iterator = 0;
+};
+
+Pseudo_LRUCacheInfo::Pseudo_LRUCacheInfo( std::size_t ways)
+    : ways( ways)
+{
+    lru_tree.insert( lru_tree.begin(), LEFT);
+    construct_tree( lru_tree.begin(), calculate_depth());
+}
+
+int Pseudo_LRUCacheInfo::calculate_depth() const
+{
+    int i = 0;
```

Comment:
You may use this function to count zero MSBs: https://github.com/MIPT-ILab/mipt-mips/blob/a28ab636c075799c0fbf9b8ad01051ef7f630dcc/simulator/infra/macro.h#L96

and by subtracting the result from bitwidth<std::size_t>, you'll get the log2 of the value (probably, it's good to extract it to a separate function)